primitiv/opencl_device.cc

```diff
@@ -98,535 +98,12 @@ std::uint32_t calc_num_blocks(std::uint32_t size, std::uint32_t num_threads) {
-  REDUCE(2)
-  REDUCE(1)
-#undef REDUCE
-  if (tid == 0) py[bid] = temp[0];
-#undef GROUP_SIZE
-}
-)EOS";
-  }
-  ss << R"EOS(
-inline float logsumexp2_fw_kernel(float a, float b) {
-  return a > b
-    ? a + log(1.f + exp(b - a))
-    : b + log(1.f + exp(a - b));
-}
-)EOS";
-  for(std::uint32_t group_size = 1; group_size <= 1024; group_size <<= 1) {
-    ss <<
-"kernel void logsumexp_fw_kernel_" << group_size << R"EOS((
-    const global float *px, const unsigned skip, const unsigned n,
-    global float *py) {
-#define GROUP_SIZE )EOS" << group_size << R"EOS(
-  const unsigned bid = get_group_id(0);
-  const unsigned tid = get_local_id(0);
-  local float temp[GROUP_SIZE];
-  px += bid % skip + (bid / skip) * skip * n;
-  temp[tid] = -1e38;
-  for (unsigned i = tid; i < n; i += GROUP_SIZE) {
-    temp[tid] = logsumexp2_fw_kernel(temp[tid], px[i * skip]);
-  }
-  barrier(CLK_LOCAL_MEM_FENCE);
-#define REDUCE(k) \
-  if (GROUP_SIZE >= k << 1) { \
-    if (tid < k) temp[tid] = logsumexp2_fw_kernel(temp[tid], temp[tid + k]); \
-    barrier(CLK_LOCAL_MEM_FENCE); \
-  }
-  REDUCE(512)
-  REDUCE(256)
-  REDUCE(128)
-  REDUCE(64)
-  REDUCE(32)
-  REDUCE(16)
-  REDUCE(8)
-  REDUCE(4)
-  REDUCE(2)
-  REDUCE(1)
-#undef REDUCE
-  if (tid == 0) py[bid] = temp[0];
-#undef GROUP_SIZE
-}
-)EOS";
-  }
-  ss << R"EOS(
-kernel void broadcast_fw_kernel(
-    const global float *px, const unsigned skip1, const unsigned skip2,
-    const unsigned size, global float *py) {
-  const unsigned i = get_global_id(0);
-  if (i < size) py[i] = px[i % skip1 + (i / skip2) * skip1];
-}
-)EOS";
-  ss << R"EOS(
-kernel void batch_sum_fw_kernel(
-    const global float *px, const unsigned size,
-    const unsigned batch, global float *py) {
-  const unsigned i = get_global_id(0);
-  if (i < size) {
-    float temp = .0f;
-    px += i;
-    for (unsigned j = 0; j < batch; ++j, px += size) {
-      temp += *px;
-    }
-    py[i] = temp;
-  }
-}
-)EOS";
-  ss << R"EOS(
-kernel void inplace_multiply_const_kernel(
-    const float k, const unsigned size, global float *px) {
-  const unsigned i = get_global_id(0);
-  if (i < size) px[i] *= k;
-}
-)EOS";
-  ss << R"EOS(
-kernel void inplace_add_kernel(
-    const global float *px, const unsigned size,
-    const unsigned mbx, const unsigned mby, global float *py) {
-  const unsigned i = get_global_id(0);
-  const unsigned bid_y = get_group_id(1);
-  const unsigned shift = bid_y * size;
-  if (i < size) atomic_add_float(py + i + mby * shift, px[i + mbx * shift]);
-}
-)EOS";
-  ss << R"EOS(
-kernel void inplace_subtract_kernel(
-    const global float *px, const unsigned size,
-    const unsigned mbx, const unsigned mby, global float *py) {
-  const unsigned i = get_global_id(0);
-  const unsigned bid_y = get_group_id(1);
-  const unsigned shift = bid_y * size;
-  if (i < size) atomic_add_float(py + i + mby * shift, -px[i + mbx * shift]);
-}
-)EOS";
-
-  return ss.str();
+const std::string generate_kernels() {
+  const std::string ss({
```

Comment:
Not necessary to use the temporary variable `ss` and the resulting string can be directly returned.